### ISSUE TITLE
Using promise library with async/await

### ISSUE THREAD

**Issue by redism** [facebook/flow#2171]
The following code does not pass flow check.

``` js
// @flow
/* global describe, it */
const Promise = require('bluebird')

describe('Flow', () => {
  it('should work for async/await pattern.', async () => {
    const value:number = await Promise.resolve(10)
    console.log(value)
  })
})
```

And here's how flow complains.

```
test/flow.spec.js:7
  7:     const value:number = await Promise.resolve(10)
                                    ^^^^^^^^^^^^^^^^^^^ call of method `resolve`
  7:     const value:number = await Promise.resolve(10)
                                    ^^^^^^^^^^^^^^^^^^^ Bluebird$Promise. This type is incompatible with
522: declare function $await<T>(p: Promise<T> | T): T;
                                   ^^^^^^^^^^^^^^ union: type application of identifier `Promise` | type parameter `T` of await. See lib: /private/tmp/flow/flowlib_d8b774b/core.js:522
  Member 1:
  522: declare function $await<T>(p: Promise<T> | T): T;
                                     ^^^^^^^^^^ type application of identifier `Promise`. See lib: /private/tmp/flow/flowlib_d8b774b/core.js:522
  Error:
    7:     const value:number = await Promise.resolve(10)
                                      ^^^^^^^^^^^^^^^^^^^ Bluebird$Promise. This type is incompatible with
  522: declare function $await<T>(p: Promise<T> | T): T;
                                     ^^^^^^^^^^ Promise. See lib: /private/tmp/flow/flowlib_d8b774b/core.js:522
  Member 2:
    7:     const value:number = await Promise.resolve(10)
                                ^^^^^^^^^^^^^^^^^^^^^^^^^ type parameter `T` of await
  Error:
    7:     const value:number = await Promise.resolve(10)
                                ^^^^^^^^^^^^^^^^^^^^^^^^^ Bluebird$Promise. This type is incompatible with
    7:     const value:number = await Promise.resolve(10)
                       ^^^^^^ number
```

Problem here seems like `await` expects native Promise type, not Bluebird$Promise. How can I use Bluebird$Promise instead of native Promise?

I'm using bluebird interface file with [this one](https://github.com/marudor/flowInterfaces/blob/master/packages/iflow-bluebird/index.js.flow).

According to this [comment](https://github.com/facebook/flow/issues/1635#issuecomment-208512377) there isn't way to do this yet. In that case I just want to raise this issue.

**Comment by avikchaudhuri** [comment]
Possible workaround for this would be to: (1) redeclare `$await`  (https://github.com/facebook/flow/blob/master/lib/core.js#L522) in a [lib] file, using an import of `'bluebird'` Promise, but importing in lib files is not supported; or (2) https://github.com/facebook/flow/issues/1635#issuecomment-208512377

**Comment by thejameskyle** [comment]
Should the definition of await be using the $Shape of Promise instead of a Promise declaration itself?

**Comment by vkurchatkin** [comment]
It seems that according to spec the only thing that is required is `then` method

**Comment by callumlocke** [comment]
@vkurchatkin is right that `await` doesn't require an ES6 Promise instance... But in fact it doesn't even require a thenable – the engine always coerces the awaited expression into a promise automatically, so you are allowed to `await` a value that **may or may not be a promise**.

The [spec defines how `await` desugars](https://tc39.github.io/ecmascript-asyncawait/#desugaring). The internal desugared logic for each awaited expression is: take the expression to the right of `await`, wrap it in `Promise.resolve(...)`, and then wait for the fulfilment value of _that_ promise.

So these are both totally valid (and are functionally identical):

```js
await Promise.resolve('foo');
await 'foo';
```

Here is a [live example](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=true&presets=stage-3&code=(async%20()%20%3D%3E%20%7B%0A%0A%20%20%2F%2F%20you%20can%20await%20a%20promise%20to%20unwrap%20it%3A%0A%20%20console.log(%20await%20Promise.resolve('FOO')%20)%3B%20%2F%2F%20'FOO'%0A%0A%20%20%2F%2F%20but%20you%20can%20also%20await%20any%20other%20value%3A%0A%20%20console.log(%20await%20'BAR'%20)%3B%20%2F%2F%20'BAR'%0A%20%20%0A%20%20%2F%2F%20and%20even%20null%2Fundefined%20values%20can%20be%20awaited%3A%0A%20%20let%20x%3B%0A%20%20console.log(%20await%20x%20)%3B%20%2F%2F%20undefined%0A%0A%7D)()%3B%0A) using Babel's desugaring, which follows the spec correctly.

Why would you ever want to `await` a non-thenable value... Possible reasons:

1. writing a library function that accepts input as either a promise or just a plain value
2. (obscure) you might want to explicitly pause the current function for the remainder of the current tick, to allow other queued events to fire before it continues.

Admittedly these aren't everyday use cases, but they are valid according to the spec, so it would be wrong for Flow to type-check awaited values as thenables.

In conclusion... Flow should allow **any** value to follow the `await` keyword, even null/undefined. (The only thing you can say for sure about typing is: the evaluated result of the entire `(await X)` expression is guaranteed *not* to be a thenable.)

**Comment by vkurchatkin** [comment]
> so it would be wrong for Flow to type-check awaited values as thenables.

No one says that await-ing non-thenable is an error, it's just what's required for a value to be unwrapped.

**Comment by callumlocke** [comment]
> No one says that await-ing non-thenable is an error, it's just what's required for a value to be unwrapped.

Not sure I follow. Can you elaborate? AFAICT, flow's core.js [declares]((https://github.com/facebook/flow/blob/25a45645f7af687cd2d073ddebb00ff5dedb1db3/lib/core.js#L521)) that `await` requires a thenable. Or am I misunderstanding?

**Comment by vkurchatkin** [comment]
No, it currently requires it to be an instance of global Promise, so it won't work with Bluebird or other libs. This PR fixes that: https://github.com/facebook/flow/pull/2291

**Comment by callumlocke** [comment]
Sorry, I meant promise. Anyway... it looks like your PR just changes `await` to also accept Thenable as well as Promise. But my point was it can be any value at all, even primitives or null/undefined. For example, `await 123` is valid. With your PR, doesn't it still fail code like `await 123`?

**Comment by vkurchatkin** [comment]
> With your PR, doesn't it still fail code like await 123?

It doesn't, and it doesn't fail now

**Comment by callumlocke** [comment]
OK looks like I misunderstood the error message

**Comment by STRML** [comment]
> Possible workaround for this would be to: (1) redeclare $await (https://github.com/facebook/flow/blob/master/lib/core.js#L522) in a [lib] file, using an import of 'bluebird' Promise, but importing in lib files is not supported; or (2) #1635 (comment)

I've had `declare function $await<T>(p: BluebirdPromise<T> | T): T;` work just fine in the definition file I'm using for Bluebird.

**Comment by barczaG** [comment]
For me the clearest workaround is to override global Promise in app entry point:
`global.Promise = require('bluebird')`
And then if you need bluebird specific methods I will import bluebird again : 
```javascript
import Bluebird from 'bluebird'
async function main () {
  Bluebird.delay(100)
}
```

**Comment by vkurchatkin** [comment]
@barczaG It won't really work, though, because Flow won't see result of `Bluebird.delay(100)` as `Promise`

**Comment by jcready** [comment]
The problem here is that an `async` function needs to be able to `await` any "thenable". [Here is an example of what should be accepted by flow](https://flowtype.org/try/#0PQKgBAAgZgNg9gdzCYAoVBjOA7AzgFzHwAsBTbAQwCMZSwBeMAbyLOwC4wAKDKgSgYA+MLy4ByEuWq0xAgL7oKuAJ7YMYKAFc1+AJY4NcONwFNUYETgJgKnAgCdd2AOYMbCCrsIAFe3AC2urikAHT2pLhwMABupOIUsuaWeIRUdviOLm4UHl5gvgFBoeGRMXGSlDSkfElYKSLpma6MOZ6EFdKkSeH4mvbYNmAA1GBUwyKockA):

```js
/* @flow */

const thenable = { then: (cb) => cb('thenable') }

async function foo () {
  const a: string = await Promise.resolve('a')
  const b: string = await Promise.resolve(thenable)
  const c: string = await thenable
  return a + b + c
}
```

Here's a basic "thenable" being `await`ed inside an `async` function on node 7 with the `--harmony-async-await` flag:

![thenable](https://i.imgur.com/0D6pelE.png)

**Comment by radex** [comment]
Hey! What's the status of this?

There was a PR fixing this: https://github.com/facebook/flow/pull/2291 — but it seems to have been reverted without explanation, and I can't find any info about the reasons why.

This is blocking me because I want to provide a `thenable` class in a _library_ — so I can't use the workaround of overriding `$await`, since it's something all users of the library would have to do.

I can send a PR again, but will be easier if I understand why the original one was rejected.

/cc @mroch @bhosmer

**Comment by SamChou19815** [comment]
Support for custom promises is not planned
